### DISCUSSION TITLE
How to use fixture/pytest_generate_tests to preprocess test data?

### DISCUSSION THREAD

**Discussion by pu-learn** [pytest-dev/pytest#11488] [Q&A]
Hello, I'm building a test framework, trying to simply the code to create a test_case.
Here's my confusion:

I have a test class, which includes multiple test cases.
And I use one csv file to save test data that need for these test cases.

I want to read csv file as setup of test class, and get test data for specific test case by test case's name according to fixture methods in conftest.py.
When csv file is read, the function return a dict, keys are name of test case, values are list of test data, and test data is saved as dict, keys are argsname of test method, values are argsvalue of the argsname.

I've tried to use two fixtures:

```
@pytest.fixture(scope="class", autouse=True)
def read_from_csv(request):
    cur_path = str(request.fspath)
    test_data = {}
    case_name_list = []
    with open(csv_path, "r", encoding="utf-8") as f:
        """ read data as dict"""
    yield test_data


@pytest.fixture(scope="function")
def get_param(request, read_from_csv):
    test_data = read_from_csv
    if test_data:
        return test_data[method_name]
```

but only get one row of test data by this method, which actually multiple rows in csv file.
So I check the document, it shows if fixture need to be executed multiple times, I should use params to transfer multiple params, but I don't know how to set my return value of fixture method to params.

Then I found pytest_generate_tests(), but it seems that hook method can't call fixutre method, so this way didn't work for me.

If there's anyway can fix my problem? Or did I get something wrong so that I failed?

**Comment by RonnyPfannschmidt** [comment]
Currently fixtures are not available at collection time
Generate-tests happens at setup time


Currently we don't document a best practice for those patterns

If feasible id recommend creating a plugin object that handles ownership of the loaded data


Unfortunately I won't be able to provide an example soon

**Reply by pu-learn** [reply]
Thanks for reply.
Since generate-tests happens at setup time, if I read csv file for a session scope fixture, can I get these data in generate-tests or other method?
